commit_rmn_ocb/validate_observation.go

```diff
@@ -0,0 +1,144 @@
+	}
+
+	if err := validateObservedOnRampMaxSeqNums(obs.OnRampMaxSeqNums, ao.Observer, observerSupportedChains); err != nil {
+		return fmt.Errorf("failed to validate OnRampMaxSeqNums: %w", err)
+	}
+
+	if err := validateObservedOffRampMaxSeqNums(obs.OffRampMaxSeqNums, ao.Observer, supportsDestChain); err != nil {
+		return fmt.Errorf("failed to validate OffRampMaxSeqNums: %w", err)
+	}
+
+	if err := commit.ValidateObservedTokenPrices(obs.TokenPrices); err != nil {
+		return fmt.Errorf("failed to validate token prices: %w", err)
+	}
+
+	if err := commit.ValidateObservedGasPrices(obs.GasPrices); err != nil {
+		return fmt.Errorf("failed to validate gas prices: %w", err)
+	}
+
+	return nil
+}
+
+func validateObservedMerkleRoots(
+	merkleRoots []MerkleRoot,
+	observer commontypes.OracleID,
+	observerSupportedChains mapset.Set[cciptypes.ChainSelector],
+) error {
+	if len(merkleRoots) == 0 {
+		return nil
+	}
+
+	seenChains := mapset.NewSet[cciptypes.ChainSelector]()
+	for _, root := range merkleRoots {
+		if !observerSupportedChains.Contains(root.ChainSel) {
+			return fmt.Errorf("found merkle root for chain %d, but this chain is not supported by Observer %d",
+				root.ChainSel, observer)
+		}
+
+		if seenChains.Contains(root.ChainSel) {
+			return fmt.Errorf("duplicate merkle root for chain %d", root.ChainSel)
+		}
+		seenChains.Add(root.ChainSel)
+	}
+
+	return nil
+}
+
+func validateObservedOnRampMaxSeqNums(
+	onRampMaxSeqNums []plugintypes.SeqNumChain,
+	observer commontypes.OracleID,
+	observerSupportedChains mapset.Set[cciptypes.ChainSelector],
+) error {
+	if len(onRampMaxSeqNums) == 0 {
+		return nil
+	}
+
+	seenChains := mapset.NewSet[cciptypes.ChainSelector]()
+	for _, seqNumChain := range onRampMaxSeqNums {
+		if !observerSupportedChains.Contains(seqNumChain.ChainSel) {
+			return fmt.Errorf("found onRampMaxSeqNum for chain %d, but this chain is not supported by Observer %d",
+				seqNumChain.ChainSel, observer)
+		}
+
+		if seenChains.Contains(seqNumChain.ChainSel) {
+			return fmt.Errorf("duplicate onRampMaxSeqNum for chain %d", seqNumChain.ChainSel)
+		}
+		seenChains.Add(seqNumChain.ChainSel)
+	}
+
+	return nil
+}
+
+func validateObservedOffRampMaxSeqNums(
+	offRampMaxSeqNums []plugintypes.SeqNumChain,
+	observer commontypes.OracleID,
+	supportsDestChain bool,
+) error {
+	if len(offRampMaxSeqNums) == 0 {
+		return nil
+	}
+
+	if !supportsDestChain {
+		return fmt.Errorf("observer %d does not support dest chain, but has observed %d offRampMaxSeqNums",
+			observer, len(offRampMaxSeqNums))
+	}
+
+	seenChains := mapset.NewSet[cciptypes.ChainSelector]()
+	for _, seqNumChain := range offRampMaxSeqNums {
+		if seenChains.Contains(seqNumChain.ChainSel) {
+			return fmt.Errorf("duplicate offRampMaxSeqNum for chain %d", seqNumChain.ChainSel)
+		}
+		seenChains.Add(seqNumChain.ChainSel)
+	}
+
+	return nil
+}
+
+func validateFChain(fChain map[cciptypes.ChainSelector]int) error {
+	for _, f := range fChain {
+		if f < 0 {
```

Comment:
Should be `f <= 0` since `f` is never 0.

Reply:
f can be 0 for chains that are not RMN enabled